File: js/6236.js
const fs = require('fs');

const inputs = fs.readFileSync('example.txt').toString().trim().split('\n');

const [N, M] = inputs[0].split(' ').map(Number);
const arr = [];
for (let i = 1; i <= N; i++) {
  arr.push(+inputs[i]);
}

let start = Math.max(...arr);
let end = arr.reduce((a, b) => a + b, 0);

while (start <= end) {
  const mid = Math.floor((start + end) / 2);

  let sum = 0;
  let cnt = 1;
  arr.forEach((e) => {
    if (sum + e > mid) {
      sum = 0;
      cnt++;
    }
    sum += e;
  });

  if (cnt <= M) {
    end = mid - 1;
  } else {
    start = mid + 1;
  }
}

console.log(start);

File: c++/6236.cpp
#include <iostream>

using namespace std;

int arr[100000];

int main() {
  freopen("example.txt", "r", stdin);
  int N, M;
  cin >> N >> M;

  int start = 0;
  int end = 0;
  for (int i = 0; i < N; i++) {
    cin >> arr[i];
    if (start < arr[i]) start = arr[i];
    end += arr[i];
  }

  while (start <= end) {
    int mid = (start + end) / 2;

    int sum = 0;
    int cnt = 1;
    for (int i = 0; i < N; i++) {
      if (sum + arr[i] > mid) {
        sum = 0;
        cnt++;
      }
      sum += arr[i];
    }

    if (cnt <= M) {
      end = mid - 1;
    } else {
      start = mid + 1;
    }
  }

  cout << start;

  return 0;
}